README.md
<h1 align="center">ibmid-login</h1>
<p>
  <a href="https://www.npmjs.com/package/ibmid-login" target="_blank">
    <img alt="Version" src="https://img.shields.io/npm/v/ibmid-login.svg">
  </a>
  <a href="#" target="_blank">
    <img alt="License: ISC" src="https://img.shields.io/badge/License-ISC-yellow.svg" />
  </a>
  <a href='https://coveralls.io/github/MarcoABCardoso/ibmid-login?branch=master'>
    <img src='https://coveralls.io/repos/github/MarcoABCardoso/ibmid-login/badge.svg?branch=master' alt='Coverage Status' />
  </a>
  <a href="#" target="_blank">
    <img alt="Node.js CI" src="https://github.com/MarcoABCardoso/ibmid-login/workflows/Node.js%20CI/badge.svg" />
  </a>
</p>

> Integrates with the IBM Identity and Account Management (IAM) and Accounts APIs to implement login using IBMid.

## Install

```sh
npm install ibmid-login
```

## Usage

As a standalone module: (see JSDoc for required parameters for each method)

```js
const { default: IBMidLogin } = require('ibmid-login')

const ibmidLogin = new IBMidLogin()
await ibmidLogin.getPasscode() // => { headers: { location: "https://identity-1.us-south.iam.cloud.ibm.com/identity/passcode" }, statusCode: 302, body: {} }
```

As an Express.js Router:

```js
const express = require('express')
const app = express()

/**
 * GET  /ibmid/passcode --> Generates an IBM Cloud one-time passcode when opened in a browser
 * POST /ibmid/login --> Send { passcode: "<PASSCODE FROM IBM CLOUD>" } to start a session (cookies)
 * POST /ibmid/logout --> Clears session (cookies)
 * GET  /ibmid/users/me --> Returns current user
 * GET  /ibmid/accounts --> Returns current user's accounts
 * GET  /ibmid/accounts --> Returns current user's accounts
 * GET  /ibmid/accounts/switch --> Send ?account_id=<NEW_ACCOUNT_ID> to switch accounts
 * GET  /ibmid/resources --> Lists resource instances
 * ALL  /ibmid/resources/:resource_id --> Resource controller API for a resource - https://cloud.ibm.com/apidocs/resource-controller/resource-controller
 * ALL  /ibmid/resources/:resource_id/<path> --> Proxy requests to the service URL.
 * e.g. If resource_id is a Watson Assistant instance, /ibmid/resources/:resource_id/v1/workspaces will proxy to the instance's /v1/workspaces endpoint.
 */
app.use('/ibmid', services.ibmid.expressAdapter)
/**
 * Authenticates all routes under /protected, redirects to /login
 */
app.use('/protected', services.ibmid.expressAdapter.authenticate({ fallback_url: '/login' }))
```

## Customization
The following options may be used to tune the behavior of this package:

- **ALLOWED_ACCOUNTS**: Array of account IDs that may be used to authenticate. Users with no access to any allowed accounts will get RC 401 for their requests.
- **ALLOWED_USERS**: Array of user emails that may be used to authenticate. Accepts regular expressions (for example, for domain allowlisting). Users whose emails do not match any of the expressions will get RC 401 for their requests.
- **IBMID_APIKEY**: API key to delegate requests to a different user when passing requests along to IBM APIs.

## Run tests

```sh
npm run test
```

## Author

👤 **Marco Cardoso**

* Github: [@MarcoABCardoso](https://github.com/MarcoABCardoso)
* LinkedIn: [@marco-cardoso](https://linkedin.com/in/marco-cardoso)

## Show your support

Give a ⭐️ if this project helped you!

CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [1.11.1](https://github.com/MarcoABCardoso/ibmid-login/compare/1.11.0...1.11.1)

> 20 September 2021

- Fix edge case in ALLOWED_ACCOUNTS [`7da479d`](https://github.com/MarcoABCardoso/ibmid-login/commit/7da479d6a8ca68137a2d6bbed1d2241113e49a4d)

#### [1.11.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.10.1...1.11.0)

> 20 September 2021

- Update authentication rules [`0093bdd`](https://github.com/MarcoABCardoso/ibmid-login/commit/0093bdd68bc998841f83188e203c8b9fafd844b1)

#### [1.10.1](https://github.com/MarcoABCardoso/ibmid-login/compare/1.10.0...1.10.1)

> 30 August 2021

- Fix unexpected response when logging in as a forbidden user [`db71273`](https://github.com/MarcoABCardoso/ibmid-login/commit/db7127384759ddd627bbee06dc42dc6d5da39a04)

#### [1.10.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.9.0...1.10.0)

> 27 August 2021

- Add ALLOWED_USERS and IBMID_APIKEY customizations [`02f88dd`](https://github.com/MarcoABCardoso/ibmid-login/commit/02f88ddfca4e1ae3a70b5c807b72b5b2f5637c1d)

#### [1.9.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.8.1...1.9.0)

> 10 August 2021

- Change speech-to-text url resolution [`496ba38`](https://github.com/MarcoABCardoso/ibmid-login/commit/496ba385d9f1c5fcaadcf7a3e5e937d21337c4b1)

#### [1.8.1](https://github.com/MarcoABCardoso/ibmid-login/compare/1.8.0...1.8.1)

> 21 July 2021

- Update docs [`f0a9b6d`](https://github.com/MarcoABCardoso/ibmid-login/commit/f0a9b6d809a71ba0e2739663b720275efd8f9510)

#### [1.8.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.7.2...1.8.0)

> 21 July 2021

- Add express adapter [`d766b02`](https://github.com/MarcoABCardoso/ibmid-login/commit/d766b020687ecb09b398a5a6b2ee699440e75cf8)
- temp [`420e3a2`](https://github.com/MarcoABCardoso/ibmid-login/commit/420e3a216f58ea91f108361820c34ee03bbc57b9)

#### [1.7.2](https://github.com/MarcoABCardoso/ibmid-login/compare/1.7.1...1.7.2)

> 19 July 2021

- Reinclude /endpoints method to API [`ce86f4a`](https://github.com/MarcoABCardoso/ibmid-login/commit/ce86f4a4e68f8b405a514985e0b026670516c948)

#### [1.7.1](https://github.com/MarcoABCardoso/ibmid-login/compare/1.7.0...1.7.1)

> 16 July 2021

- Fix uncaught error when providing an invalid refresh token [`f8ef7e6`](https://github.com/MarcoABCardoso/ibmid-login/commit/f8ef7e6a231fe5e3f603abfa1aa5a95c027b8bcc)

#### [1.7.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.6.0...1.7.0)

> 16 July 2021

- Add JWT validation and split getOwnUser and listAccounts methods [`62261c7`](https://github.com/MarcoABCardoso/ibmid-login/commit/62261c7bab60928f92665b59b0a1510ff9917bea)

#### [1.6.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.5.0...1.6.0)

> 13 July 2021

- Use /watson_data path for Watson Data API resources [`77b7ff4`](https://github.com/MarcoABCardoso/ibmid-login/commit/77b7ff48a372c941614b08a85905ac29ed64c7a0)

#### [1.5.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.4.0...1.5.0)

> 12 July 2021

- Add apikey as an authentication method [`c9c3010`](https://github.com/MarcoABCardoso/ibmid-login/commit/c9c30109ccc1a4b390abbb65f2abc1a4ae5ed541)

#### [1.4.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.3.0...1.4.0)

> 30 June 2021

- Add manageResource to the API [`a949583`](https://github.com/MarcoABCardoso/ibmid-login/commit/a94958356eed28c9570df69750d5a7430e45fe18)

#### [1.3.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.2.0...1.3.0)

> 28 June 2021

- Add proxy and listResources methods to IBMidService [`e4e54ba`](https://github.com/MarcoABCardoso/ibmid-login/commit/e4e54baf811b51ebeb78d0044d6397ad604200c8)

#### [1.2.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.1.1...1.2.0)

> 25 June 2021

- Add support for TS default import [`6631357`](https://github.com/MarcoABCardoso/ibmid-login/commit/6631357bb38b7aee165d63838cea42810d3a7495)

#### [1.1.1](https://github.com/MarcoABCardoso/ibmid-login/compare/1.1.0...1.1.1)

> 25 June 2021

- Update authenticatedFunction interface [`94e44e2`](https://github.com/MarcoABCardoso/ibmid-login/commit/94e44e2566420399b74ed67e2abe057718176ecd)

#### [1.1.0](https://github.com/MarcoABCardoso/ibmid-login/compare/1.0.3...1.1.0)

> 25 June 2021

- Change exported module to class [`4ab9045`](https://github.com/MarcoABCardoso/ibmid-login/commit/4ab9045582d485c7487360c26d7d82d540502ce8)

#### [1.0.3](https://github.com/MarcoABCardoso/ibmid-login/compare/1.0.2...1.0.3)

> 25 June 2021

- Fix and add test to default instance construction [`8a0cb63`](https://github.com/MarcoABCardoso/ibmid-login/commit/8a0cb6365c29e3de9b853a4497c47247119eb4fc)

#### [1.0.2](https://github.com/MarcoABCardoso/ibmid-login/compare/1.0.1...1.0.2)

> 25 June 2021

- Move @types/jest to devDependencies [`61c610a`](https://github.com/MarcoABCardoso/ibmid-login/commit/61c610a50dfa81746212974b26bc4a1170d8a877)

#### [1.0.1](https://github.com/MarcoABCardoso/ibmid-login/compare/1.0.0...1.0.1)

> 24 June 2021

- Remove -g flag from README install instructions [`a680e66`](https://github.com/MarcoABCardoso/ibmid-login/commit/a680e6655a627691f080f11fc945ab1334267a0d)

### [1.0.0](https://github.com/MarcoABCardoso/ibmid-login/compare/0.0.1...1.0.0)

> 24 June 2021

- Remove need for static fields [`24c4456`](https://github.com/MarcoABCardoso/ibmid-login/commit/24c4456f7ddbcf8cfc9f61f0e077187c95fe4929)

#### 0.0.1

> 24 June 2021

- Initial commit [`347427d`](https://github.com/MarcoABCardoso/ibmid-login/commit/347427d5d0d16c50659f2ea12049a90a2a1b4ef7)


lib/ibmid-service.js
const axios = require('axios')
const AuthenticatedService = require('./internal/authenticated-service')
const IAMAPI = require('./internal/iam-api')
const AccountsAPI = require('./internal/accounts-api')
const GlobalCatalogAPI = require('./internal/global-catalog-api')
const ResourceControllerAPI = require('./internal/resource-controller-api')
const { notLoggedInResponse } = require('./responses')
const getExpressAdapter = require('./adapters/express')
const constants = require('./config/constants')
const memoizee = require('memoizee')


class IBMidService extends AuthenticatedService {

    constructor(
        options,
        requestApi = axios,
        iamApi = new IAMAPI(requestApi),
        accountsApi = new AccountsAPI(requestApi),
        globalCatalogApi = new GlobalCatalogAPI(requestApi),
        resourceControllerApi = new ResourceControllerAPI(requestApi)
    ) {
        super(requestApi, iamApi)
        this.options = { ...constants, ...options }
        this.accountsApi = accountsApi
        this.globalCatalogApi = globalCatalogApi
        this.resourceControllerApi = resourceControllerApi
        this.getPasscode = this.getPasscode.bind(this)
        this.login = this.login.bind(this)
        this.memoizedLogin = memoizee(this.login.bind(this), { maxAge: this.options.IBMID_APIKEY_LOGIN_CACHE_TIMEOUT, normalizer: args => JSON.stringify(args) })
        this.logout = this.logout.bind(this)
        this.switchAccount = this.authenticatedFunction(this.switchAccount.bind(this))
        this.getOwnUser = this.authenticatedFunction(this.getOwnUser.bind(this))
        this.listAccounts = this.authenticatedFunction(this.listAccounts.bind(this))

        if (this.options.IBMID_APIKEY) {
            this.proxy = this.withTokenOverride(this.proxy.bind(this))
            this.listResources = this.withTokenOverride(this.listResources.bind(this))
            this.manageResource = this.withTokenOverride(this.manageResource.bind(this))
        }

        this.proxy = this.authenticatedFunction(this.proxy.bind(this))
        this.listResources = this.authenticatedFunction(this.listResources.bind(this))
        this.manageResource = this.authenticatedFunction(this.manageResource.bind(this))

        this.expressAdapter = getExpressAdapter(this)
    }

    async getPasscode() {
        return this.iamApi.getOpenIDConfig()
            .then(data => ({ headers: { location: data.passcode_endpoint }, statusCode: 302, body: {} }))
    }

    async login({ apikey, passcode }) {
        let data = await this.iamApi.createIAMToken({ apikey, passcode })
        if (!data.success) return { statusCode: 401, body: data }
        if (apikey) return { statusCode: 200, body: data, headers: this.getAuthCookies({ token: data.token, refreshToken: data.refresh_token, expiresIn: data.expires_in }) }
        let accountData = await this.listAccounts({ token: data.token })
        let account_id, refreshData
        for (let account of accountData.body.resources) {
            if (refreshData && refreshData.success) continue
            account_id = account.metadata.guid
            refreshData = await this.iamApi.createIAMToken({ refreshToken: data.refresh_token, accountID: account_id })
        }
        if (refreshData && refreshData.success) data = refreshData
        let userData = await this.getOwnUser({ token: data.token })
        if (userData.statusCode >= 400) return notLoggedInResponse()
        let headers = this.getAuthCookies({ token: data.token, refreshToken: data.refresh_token, accountID: account_id, expiresIn: data.expires_in })
        return { statusCode: 200, body: data, headers }
    }

    async logout() {
        let headers = this.getAuthCookies({})
        return { statusCode: 200, body: { success: true }, headers }
    }

    async switchAccount({ refreshToken, accountID }) {
        if (accountID && this.options.ALLOWED_ACCOUNTS && !this.options.ALLOWED_ACCOUNTS.includes(accountID))
            return notLoggedInResponse()
        let refreshTokenResponse = await this.refreshToken({ refreshToken, accountID })
        return { statusCode: 200, ...refreshTokenResponse }
    }

    async getOwnUser({ token }) {
        return this.iamApi.validateToken({ token })
            .then(user => {
                if (this.checkUserAllowed(user)) return { statusCode: 200, body: user }
                throw new Error('User is not allowed')
            })
            .catch(() => notLoggedInResponse())
    }

    checkUserAllowed(user) {
        return (!this.options.ALLOWED_USERS && !this.options.ALLOWED_ACCOUNTS) ||
            (this.options.ALLOWED_USERS && this.options.ALLOWED_USERS.reduce((acc, allowed) => acc || !!user.email.match(allowed), false)) ||
            (this.options.ALLOWED_ACCOUNTS && user.account && this.options.ALLOWED_ACCOUNTS.includes(user.account.bss))
    }

    async listAccounts({ token }) {
        return this.accountsApi.listAccounts({ token })
            .then(response => ({
                ...response.data,
                resources: response.data.resources.filter(acc => !this.options.ALLOWED_ACCOUNTS || this.options.ALLOWED_ACCOUNTS.includes(acc.metadata.guid))
            }))
            .then(data => ({ statusCode: 200, body: data }))
            .catch(err => ({ statusCode: err.response.status, body: err.response.data }))
    }

    async listResources({ token, resourceType }) {
        let resourceID = null
        if (resourceType) {
            let catalogListData = await this.globalCatalogApi.listCatalogEntries({ resourceType })
            let catalogEntry = catalogListData.resources
                .reduce((acc, r) => r.children ? [...acc, ...r.children] : acc, catalogListData.resources)
                .find(r => r.kind === 'service' && r.name === resourceType)
            if (!catalogEntry) return { statusCode: 404, body: { message: 'Resource not found' } }
            resourceID = catalogEntry.id
        }

        let resourceListData = await this.resourceControllerApi.listAllResources({ token, resourceID })
        return {
            statusCode: 200,
            body: resourceListData
        }

    }

    async manageResource({ token, resourceID, url, method, params, data }) {
        return this.resourceControllerApi.manageResource({ token, resourceID, url, method, params, data })
    }

    proxyRequest(targetURL) {
        let self = this
        return function ({ token, url, method, headers, params, data }) {
            let iamOptions = { url, method, headers, params, data }
            delete iamOptions.headers.host
            iamOptions.headers.authorization = `Bearer ${token}`
            if (url.includes('daas')) delete iamOptions.headers
            iamOptions.url = `${targetURL}${url}`
            return self.requestApi(iamOptions)
                .then(response => ({ statusCode: response.status, body: response.data, headers: response.headers }))
                .catch(err => ({ statusCode: err.response.status, body: err.response.data, headers: err.response.headers }))
        }
    }

    async getProxyURL({ token, resourceID, url, headers }) {
        if (resourceID.includes('whisk.system')) return `https://${resourceID.match(/(.*)functions:(.*):a/)[2]}.functions.cloud.ibm.com/api/v1/namespaces/${decodeURIComponent(resourceID).match(/(.*):(.*)::/)[2]}`
        if (resourceID == 'watson_data') return 'https://api.dataplatform.cloud.ibm.com'
        let resourceResponse = await this.manageResource({ token, resourceID, url: '', method: 'GET' })
        if (resourceResponse.statusCode >= 400 || resourceResponse.body.status_code >= 400) return
        let resource = resourceResponse.body
        if (resource.resource_id === 'functions')
            return `https://${resource.region_id}.functions.cloud.ibm.com/api/v1/${url.startsWith('/web/') ? 'web' : 'namespaces'}/${resource.guid.match(/(.*):(.*)::/)[2]}${url.startsWith('/web/') ? url.replace('/web', '') : ''}`
        if (resource.id.includes('conversation'))
            return `https://api.${resource.region_id}.assistant.watson.cloud.ibm.com/instances/${resource.guid}`
        if (resource.id.includes('speech-to-text'))
            return `https://api.${resource.region_id}.speech-to-text.watson.cloud.ibm.com/instances/${resource.guid}`
        let keys = await this.manageResource({ token, resourceID, url: '/resource_keys', method: 'GET' }).then(response => response.body)
        let managerCreds = keys.resources.find(key => key.role && key.role.includes('Manager'))
        let editorCreds = keys.resources.find(key => key.role && key.role.includes('Writer'))
        let readerCreds = keys.resources.find(key => key.role && key.role.includes('Reader'))
        let bestCreds = managerCreds || editorCreds || readerCreds || {}
        let credentials = bestCreds.credentials || {}
        if (credentials.url)
            return credentials.url
        if (credentials.api_endpoint_url)
            return credentials.api_endpoint_url.replace('https://', `https://${credentials.client_id}:${credentials.client_secret}@`).replace('/daas/', '')
        if (credentials.endpoints) {
            if (url === '/endpoints') return credentials.endpoints.replace('/endpoints', '')
            if (headers['x-endpoint-id']) {
                let endpoints = await this.requestApi({ url: credentials.endpoints }).then(response => response.data)
                let endpointPath = headers['x-endpoint-id'].split(':')
                return 'https://' + endpointPath.reduce((endpoint, step) => endpoint[step], endpoints)
            }
        }
    }

    async proxy({ token, resourceID, url, method, headers = {}, params, data }) {
        let proxyURL = await this.getProxyURL({ token, resourceID, url, headers })
        if (!proxyURL) return { statusCode: 404, body: { message: 'Resource not found' } }
        return this.proxyRequest(proxyURL)({ token, url, method, headers, params, data })
    }

    withTokenOverride(func) {
        return async options => {
            let loginResponse = await this.memoizedLogin({ apikey: this.options.IBMID_APIKEY })
            if (loginResponse.statusCode >= 400) return func(options)
            return func({ ...options, token: loginResponse.body.token })
        }
    }

}

IBMidService.default = new IBMidService()

module.exports = IBMidService